R/annotate.R

```diff
@@ -0,0 +1,120 @@
+#' sign is added before month in date (e.g. 1916-?10-10 or 1916-~10-10).
+#' @return A messydt object with annotated date(s)
+#' @examples
+#' data <- data.frame(Beg = c("1816-01-01", "1916-01-01", "2016-01-01"),
+#' End = c("1816-12-31", "1916-12-31", "2016-12-31"))
+#' dplyr::mutate(data, Beg = ifelse(Beg <= "1816-01-01",
+#'  on_or_before(Beg), Beg))
+#' dplyr::mutate(data, End = ifelse(End >= "2016-01-01",
+#'  on_or_after(End), End))
+#' dplyr::mutate(data, Beg = ifelse(Beg == "1916-01-01",
+#' add_approximation(Beg), Beg))
+#' dplyr::mutate(data, End = ifelse(End == "1916-12-31",
+#' add_uncertainty(End), End))
+#' @name annotate
+NULL
+
+#' @rdname annotate
+#' @details `on_or_before()` annotates uncertain start dates by adding ".." as
+#' a prefix to dates (e.g. `..1816-01-01`).
+#' @export
+on_or_before <- function(x) {
+x <- paste0("..", x)
+x <- as_messydate(x)
+x
+}
+
+#' @rdname annotate
+#' @details `on_or_after()` annotates uncertain end dates by adding ".." as
+#' a suffix to dates (e.g. `2016-12-31`).
+#' @export
+on_or_after <- function(x) {
+  x <- paste0(x, "..")
+  x <- as_messydate(x)
+  x
+}
+
+#' @rdname annotate
+#' @details `add_approximation()` annotates approximate dates, or date
+#' components, deemed possibly correct by adding "~" to date
+#' (e.g. `1916~-01-01`)
+#' @export
+add_approximation <- function(x, level = NULL) {
+
+  if (is.null(level)) {
+    x <- paste0("~", x)
+  }
+
+  if (!is.null(level)) {
+    day <- vapply(strsplit(x, "-"), `[`, 3, FUN.VALUE = character(1))
+    month <- vapply(strsplit(x, "-"), `[`, 2, FUN.VALUE = character(1))
+    year <- vapply(strsplit(x, "-"), `[`, 1, FUN.VALUE = character(1))
+    if (level == "day") {
+      x <- paste0(year, "-", month, "-", day, "~")
+    }
+    if (level == "month") {
+      x <- paste0(year, "-", month, "~", "-", day)
+    }
+    if (level == "year") {
+      x <- paste0(year, "~", "-", month, "-", day)
+    }
+    if (level == "md") {
+      x <- paste0(year, "-", "~", month, "-", day)
+    }
+  }
+  x <- as_messydate(x)
+  x
+}
+
+#' @rdname annotate
+#' @details `add_uncertainty()` annotates uncertain dates, or date components,
+#' deemed dubious by adding "?" to date (e.g. `1916?-01-01`)
+#' @export
+add_uncertainty <- function(x, level = NULL) {
+
+  if (is.null(level)) {
+    x <- paste0("?", x)
+  }
+
+  if (!is.null(level)) {
+    day <- vapply(strsplit(x, "-"), `[`, 3, FUN.VALUE = character(1))
+    month <- vapply(strsplit(x, "-"), `[`, 2, FUN.VALUE = character(1))
+    year <- vapply(strsplit(x, "-"), `[`, 1, FUN.VALUE = character(1))
+    if (level == "day") {
+      x <- paste0(year, "-", month, "-", day, "?")
+    }
+    if (level == "month") {
+      x <- paste0(year, "-", month, "?", "-", day)
+    }
+    if (level == "year") {
+      x <- paste0(year, "?", "-", month, "-", day)
+    }
+    if (level == "md") {
+      x <- paste0(year, "-", "?", month, "-", day)
+    }
+  }
+  x <- as_messydate(x)
+  x
+}
```

Comment:
I wonder whether we are really _adding_ uncertainty/approximation, or are we just _annotating_ a date or character string with this information?

Reply:
@jhollway @henriquesposito could change it to annotate_uncertainty/approximation/etc, although I think the current 'add' implies add annotation

Reply:
`as_approximate()`/`as_uncertain()`?

Reply:
Yes, agreed, "as_" prefix sounds better and is more consistent!